Feature/4105 sharing detection

aerych opened the pull request:
This PR improves how support for Sharing (via Publicize and share buttons) is detected on wpcom and Jetpack connected blogs.  The sharing feature is always on for wpcom blogs, but is either on or off for Jetpack connected blogs depending on which modules are enabled.  The testing scenarios should cover most those situations.

Some work was done to make sure options were always up to date before viewing a blog's details screen. I'm a bit iffy on `BlogSyncState` as a way to get a completion handler for `syncBlog:`, but it seemed a bit cleaner than breakout out the options sync, or implementing multiple call backs. Very open to other ideas.

Needs review: @koke 

### Testing
Scenario 1:
Sign into with a wpcom account and view the blog details screen of a wpcom blog.  Sharing should be present.

Scenario 2: 
Sign into a wpcom account that has a Jetpack connected blog. 
View that the details screen for the Jetpack connected blog.  Sharing should be present.

Scenario 3:
Be signed out of wpcom
Sign into a self-hosted blog and view its blog details screen. Sharing should *not* be present

Scenario 4:
Repeat the steps in Scenario 3.
Tap on the Stats feature. Sign in to Jetpack when prompted.
Return to the details screen.  Sharing should be present. 

Scenario 5: 
Repeat the steps in Scenario 3.
Go to the Me tab and sign into the wpcom account that is connected to the self-hosted site's Jetpack. 
Return to the My Sites tab and view the detail for the self-hosted blog. Sharing should be present. 

Scenario 6: 
Be signed into wpcom. 
Sign into a Jetpack connected blog that is connected to a *different* wpcom account. 
Go to that blog's details screen. Sharing should *not* be present. 
Tap on Stats and connect to Jetpack when prompted. 
Return to the details screen.  Sharing should be present. 

Scenario 7: 
Disable both Publicize and Sharing buttons on a Jetpack connected blog. 
Repeat the steps in Scenario 2.  Sharing should *not* be present.

Scenario 8: 
Disable either Publicize or Sharing buttons on a Jetpack connected blog. 
Repeat the steps in Scenario 2.  Sharing should be present.  Note that the Sharing VC does not yet reflect one of the features being enabled and one not. This will be handled in a future PR. 

Scenario 9: 
Disable the JSON API module on a Jetpack connected blog. 
Repeat the steps in Scenario 2.  Sharing should *not* be present.

koke commented on WordPress/Classes/Models/Blog.m:
```diff
@@ -439,6 +444,46 @@ - (BOOL)supports:(BlogFeature)feature
+    // wpcom blog and publicize has not been disabled
+    if (self.isHostedAtWPcom && ![[self getOptionValue:OptionsKeyPublicizeDisabled] boolValue]) {
+        return YES;
+    }
```
I think if `isHostedAtWPcom` is true but publicize is disabled, it should return `NO` immediately.

aerych replied:
:+1:

koke commented on WordPress/Classes/Models/Blog.m:
```diff
@@ -439,6 +444,46 @@ - (BOOL)supports:(BlogFeature)feature
+    // self-hosted connected to jetpack via the wpcom acct signed into the app.
+    if (self.dotComID && [self jetpackPublicizeModuleEnabled]) {
+        return YES;
+    }
```
I think this `if` block is unnecessary. `jetpackRESTSupported` below already handles this case

aerych replied:
`jetpackRESTSupported` returns `NO` in this case as `self.jetpackAccount` is nil.

koke replied:
I see, I forgot the first check was `HostedAtWPcom` and not just REST. I still find the logic a bit hard to follow. How about this?

```obj-c
    // Publicize is only supported via REST
    if (![self supports:BlogFeatureWPComRESTAPI]) {
        return NO;
    }

    if (self.isHostedAtWPcom) {
        // For WordPress.com YES unless it's disabled
        return ![[self getOptionValue:OptionsKeyPublicizeDisabled] boolValue];
    } else {
        // For Jetpack, check if the module is enabled
        return [self jetpackPublicizeModuleEnabled];
    }
```

aerych replied:
Doesn't seem hard to follow to me, but then I wrote it. ;P
Sure. That works just as well so I'll make the change.

koke commented on WordPress/Classes/Models/Blog.m:
```diff
@@ -439,6 +444,46 @@ - (BOOL)supports:(BlogFeature)feature
+    // self-hosted connected to jetpack via the wpcom acct signed into the app.
+    if (self.dotComID && [self jetpackSharingButtonsModuleEnabled]) {
+        return YES;
+    }
```
Also unnecessary, like the one in `supportsPublicize`

aerych replied:
see above :)

aerych replied:
Fwiw, I considered consolidating the two if blocks into something like 
`if ((self.dotComID || [self jetpackRESTSupported]) && [self jetpackSharingButtonsModleEnabled)` but I decided separating the checks was better for clarity of purpose.

koke commented:
> I'm a bit iffy on BlogSyncState

Agreed, it smells like trouble ahed :wink:
Some issues I can detect:
- If you add/remove requests but forget to update the request count :boom: 
- Not thread safe
- We aggregate all the errors, but then nothing uses them

Some other ideas:
- Batch requests: they are problematic on self hosted, so no :stuck_out_tongue: 
- NSOperation dependencies: get an operation for each request, then have a syncBlog operation that depends on the previous 4. See `-[CreateAccountAndBlogViewController createUserAndSite]`. Requires a bit of refactor, but the best solution IMO.
- Use [dispatch_group](http://commandshift.co.uk/blog/2014/03/19/using-dispatch-groups-to-wait-for-multiple-web-services/). Doesn't require changes on the remotes or the categories service.
- If you prefer BlogSyncState, use the [OSAtomic](https://developer.apple.com/library/mac/documentation/System/Reference/OSAtomic_header_reference/#//apple_ref/c/func/OSAtomicDecrement64Barrier) functions to check and modify the counter.

koke commented on WordPress/Classes/Networking/RemoteBlogOptionsHelper.m:
```diff
@@ -0,0 +1,47 @@
+        }
+    } else {
+        //valid default values
+        options[@"software_version"] = @"3.6";
```
I can almost see the puzzled look on some poor user's face when the app asks them to upgrade the WordPress version in WordPress.com. I wonder if `response[@"options"] == nil` should be treated as an error

aerych commented:
> Some other ideas

I really liked the `NSOperation` approach, but it would need the XMLRPC remotes to return NSOperations and that would require a pod change.  That's a bit more of a refactor than I'm looking to do here. 

Instead I've gone with the `dispatch_group` approach. Thanks for the tip!!  I love how light weight the change is.

Ready for another round @koke :D

koke commented on WordPress/Classes/Models/Blog.m:
```diff
@@ -20,6 +20,11 @@
+NSString * const ActiveModulesKeyPublicize = @"publicize";
+NSString * const ActiveModulesKeySharingButtons = @"sharedaddy";
+NSString * const OptionsKeyActiveModules = @"active_modules";
+NSString * const OptionsKeyPublicizeDisabled = @"publicize-permanently-disabled";
```
I think this should be `publicize_permanently_disabled`

koke commented:
On a Jetpack site, with Publicize disabled and Sharing enabled, I get this:

![simulator screen shot 02 feb 2016 08 42 29](https://cloud.githubusercontent.com/assets/8739/12742789/0ce5b46a-c989-11e5-99cd-51eb03e416bd.png)

With Publicise enabled and Sharing disabled there's no error. Although both sections are visible if one of the modules is disabled.

Then, possibly out of scope, Facebook connection fails. Or any other connection. But now it fixed itself and I can't reproduce :angry: 

![simulator screen shot 02 feb 2016 08 45 53](https://cloud.githubusercontent.com/assets/8739/12742827/6561c0b6-c989-11e5-859a-df8eef53d36d.png)
![simulator screen shot 02 feb 2016 08 45 56](https://cloud.githubusercontent.com/assets/8739/12742828/658aa1d4-c989-11e5-9408-fcaf019d71c1.png)

koke commented:
Just to leave a note, and related to #3761, I believe "Sharing" should not hide if Jetpack is not configured but, like Stats, it should be a prompt to configure Jetpack. It's kind of weird and unexpected that you have to go through stats to enable a different Jetpack feature.

koke commented on WordPress/Classes/ViewRelated/Stats/JetpackSettingsViewController.m:
```diff
@@ -457,11 +457,15 @@ - (IBAction)saveAction:(id)sender
+        // Ensure options are up to date after connecting Jetpack as there may
+        // now be new info.
+        BlogService *service = [[BlogService alloc] initWithManagedObjectContext:[[ContextManager sharedInstance] mainContext]];
+        [service syncBlog:self.blog completionHandler:^(NSArray *failures) {
```
This one is crashing because of the extra array parameter

aerych replied:
Oops. Missed that one when I refactored.

koke commented:
This one's weird. I logged in with a test wp.com account (kokejb7), then added a Jetpack site (wpt.koke.me) connected to a different account (kokeme), went to stats and connected Jetpack. Both Publicize and Sharing are enabled. The Jetpack site only has one connected account, and it's an admin. However I get this:

![simulator screen shot 02 feb 2016 09 11 58](https://cloud.githubusercontent.com/assets/8739/12743312/72e4b9c4-c98d-11e5-8ebf-1d0100c93034.png)

And when I try to connect:

![simulator screen shot 02 feb 2016 09 11 45](https://cloud.githubusercontent.com/assets/8739/12743321/7d326502-c98d-11e5-8326-d68c277aa090.png)

If I break on the publicize sync failure and inspect the error:

```
Error Domain=com.wordpress.api Code=4 "This user is not authorized to access publicize connections on this site. [authorization_required]" UserInfo={WordPressComApiErrorCodeKey=authorization_required, NSLocalizedDescription=This user is not authorized to access publicize connections on this site. [authorization_required]}
```

koke commented:
> Debugging it more, it seems it's using the wrong token (kokejb7)

Indeed, `SharingService.apiForRequest()` is always using the api from the default account, regardless of blog.

aerych commented:
> On a Jetpack site, with Publicize disabled and Sharing enabled ...

Yep!  Much of the UI is still a work in progress. The sharing vc will be updated later to show/hide supported features in a subsequent PR. 

>  I believe "Sharing" should not hide if Jetpack is not configured but, like Stats, it should be a prompt to configure Jetpack

Agreed, and it should also include instructions on how to enable the relevant disabled Jetpack modules. Very much related to the note above. 

> Indeed, `SharingService.apiForRequest()` is always using the api from the default account, regardless of blog

DoH! Great catch right there. I've changed things so the blog's `restApi` is used. I also fixed some issues authenticating Jetpack connected blogs when making the oAuth hand shake. Maybe a bit out of scope for this PR but oh well.  

Ready for another round @koke :)

koke commented on WordPress/Classes/Services/SharingService.swift:
```diff
@@ -451,25 +455,20 @@ public class SharingService : LocalCoreDataService
 
-        return api!
+        DDLogSwift.logError("SharingService is not using a real WordPress.com account.")
+        return WordPressComApi.anonymousApi()
```
Did you drop the `assert` intentionally? Also `assert(false) == assertionFailure()` :)

aerych replied:
Yep.  The assert doesn't make sense when we're intending to return the anonymous api. However, that `logError` should probably be a `logDebug` with a better message.

koke commented on WordPress/Classes/Services/SharingService.swift:
```diff
@@ -451,25 +455,14 @@ public class SharingService : LocalCoreDataService
+    private func apiForBlog(blog: Blog) -> WordPressComApi {
+        let api: WordPressComApi? =  blog.restApi()
+        assert(api != nil)
         return api!
```
We talked about it on Slack, but leaving a note here, for future reference. I don't like the forced unwrap and style of this, but it's good for now. When we add nullability annotations to `Blog` (or convert it to swift :wink: ) we'll be able to avoid the cast to optional.

As for the force unwrap, I think it'd require passing the blog when initializing, so we can fail if blog.restApi is nil. But even then, the calling code would have to handle the possibility of a nil service.

Anyway, good to go, just wanted to write my thoughts somewhere :smile:

aerych replied:
> we add nullability annotations to Blog [...] we'll be able to avoid the cast to optional.

I'll happily remove the optional cast when that happens!  For now I think the extra clarity the deliberate unwrapping provides is a good thing.  Thanks again :)

koke commented:
:shipit:

aerych commented:
Thanks Koke!!
